cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -635,27 +668,50 @@ public void onClose(KubernetesClientException e) {
                                                                  boolean dryrun, RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder) {
+        return requestRebalance(reconciliation, host, apiClient, dryrun, rebalanceOptionsBuilder, null);
+    }
+
+    private Future<KafkaClusterRebalanceStatus> requestRebalance(Reconciliation reconciliation, String host, CruiseControlApi apiClient,
+        boolean dryrun, RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder, String userTaskID) {
+
         log.info("{}: Requesting Cruise Control rebalance [dryrun={}]", reconciliation, dryrun);
         if (!dryrun) {
             rebalanceOptionsBuilder.withFullRun();
         }
-        State ready = dryrun ? State.ProposalReady : State.Ready;
-        State inprogress = dryrun ? State.PendingProposal : State.Rebalancing;
-        return apiClient.rebalance(host, CruiseControl.REST_API_PORT, rebalanceOptionsBuilder.build())
+        return apiClient.rebalance(host, CruiseControl.REST_API_PORT, rebalanceOptionsBuilder.build(), userTaskID)
                 .map(response -> {
-                    if (response.thereIsNotEnoughDataForProposal()) {
-                        return new KafkaClusterRebalanceStatusBuilder()
-                                .withNewSessionId(response.getUserTaskId())
-                                .addNewCondition().withNewType(inprogress.toString()).endCondition().build();
+                    if (dryrun) {
+                        if (response.thereIsNotEnoughDataForProposal() || response.proposalIsStillCalculating()) {
+                            // If either there not enough data for a rebalance or it is still being processed we need to
+                            // re-request the proposal at a later stage so we move to the PendingProposal State.
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
+                        }
                     } else {
-                        if (response.getJson().containsKey(CC_REST_API_SUMMARY)) {
+                        if (response.thereIsNotEnoughDataForProposal()) {
                             return new KafkaClusterRebalanceStatusBuilder()
-                                    .withOptimizationResult(response.getJson().getJsonObject(CC_REST_API_SUMMARY).getMap())
-                                    .addNewCondition().withNewType(ready.toString()).endCondition().build();
-                        } else {
-                            throw new RuntimeException("Rebalance returned unknown response: " + response.toString());
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
```

Comment:
so we are in `!dryrun` branch and it's possible that the state machine moves to `PendingProposal`. It sounds to me that we are introducing a new potential transition from `Rebalancing` to `PendingProposal` which sounds a little bit weird. Can you enlight me more about this branch?

Reply:
We cannot move from `Rebalancing` to `PendingProposal` as the `onRebalancing` method does not call this method it calls the `user_tasks` endpoint instead. The only way you hit this branch is if you do a `rebalance(dryrun=false)` and the proposal is still calculating. In which case you will be in an active rebalance when that proposal completes (with no user or CO intervention) so we move to `Rebalancing`.